File: Common/HashCodeComparer.vb
Public Class HashCodeComparer(Of T)
    Implements IComparer, IComparer(Of T), IEqualityComparer, IEqualityComparer(Of T)

    Public Function Compare(x As T, y As T) As Integer Implements IComparer(Of T).Compare
        Return x.GetHashCode().CompareTo(y.GetHashCode())
    End Function

    Public Overloads Function GetHashCode(obj As T) As Integer Implements IEqualityComparer(Of T).GetHashCode
        Return obj.GetHashCode()
    End Function

    Public Overloads Function Equals(x As T, y As T) As Boolean Implements IEqualityComparer(Of T).Equals
        Return x.GetHashCode() = y.GetHashCode()
    End Function

    Private Function Compare(x As Object, y As Object) As Integer Implements IComparer.Compare
        Return Me.Compare(DirectCast(x, T), DirectCast(y, T))
    End Function

    Private Overloads Function GetHashCode(obj As Object) As Integer Implements IEqualityComparer.GetHashCode
        Return Me.GetHashCode(DirectCast(obj, T))
    End Function

    Private Overloads Function Equals(x As Object, y As Object) As Boolean Implements IEqualityComparer.Equals
        Dim TX = DirectCast(x, T)
        Dim TY = DirectCast(y, T)

        If TX Is Nothing Or TY Is Nothing Then
            Return x Is Nothing And y Is Nothing
        End If

        Return Me.Equals(TX, TY)
    End Function

End Class

File: Common/HashCodeComparer.cs
using System.Collections;
using System.Collections.Generic;

namespace Ks
{
    public class HashCodeComparer<T> : IComparer, IComparer<T>, IEqualityComparer, IEqualityComparer<T>
    {
        public int Compare(T x, T y)
        {
            return x.GetHashCode().CompareTo(y.GetHashCode());
        }

        public int GetHashCode(T obj)
        {
            return obj.GetHashCode();
        }

        public bool Equals(T x, T y)
        {
            return x.GetHashCode() == y.GetHashCode();
        }

        int IComparer.Compare(object x, object y)
        {
            return this.Compare((T) x, (T) y);
        }

        int IEqualityComparer.GetHashCode(object obj)
        {
            return this.GetHashCode((T) obj);
        }

        bool IEqualityComparer.Equals(object x, object y)
        {
            var TX = (T) x;
            var TY = (T) y;

            if (TX == null | TY == null)
            {
                return x == null & y == null;
            }

            return this.Equals(TX, TY);
        }
    }
}
